src/adapter/typename.rs

```diff
@@ -0,0 +1,448 @@
+use std::fmt::{Display, Formatter, Result};
+
+macro_rules! display_wrapper {
+    ($vis:vis $t:ident,
+    $impl:item) => {
+        #[derive(Debug, Clone, Copy)]
+        $vis struct $t<'a>($vis &'a rustdoc_types::$t);
+
+        impl Display for $t<'_> {
+            $impl
+        }
+    };
+}
+
+fn intersperse<T, I: IntoIterator<Item = T>, F: FnMut(T, &mut Formatter<'_>) -> Result>(
+    fmt: &mut Formatter<'_>,
+    delim: &str,
+    items: I,
+    mut f: F,
+) -> Result {
+    let mut first = true;
+    for item in items {
+        if first {
+            first = false;
+        } else {
+            write!(fmt, "{delim}")?;
+        }
+
+        f(item, fmt)?;
+    }
+
+    Ok(())
+}
+
+#[inline]
```

Comment:
Hm, I didn't know that.  I just didn't want the overhead of a function call here, but LLVM may figure that out on its own, since we don't want to inline the `intersperse` call.